### ISSUE TITLE
For Mac, have the option to use FUSE-T

### ISSUE THREAD

**Issue by dfloyd888** [veracrypt/VeraCrypt#1055]
# Desired behavior
Please add, as an option for macOS, the ability to use FUSE-T.  FUSE-T is similar to FUSE, but instead of presenting volumes directly to the OS, it presents them as loopback mounted NFS v4 volumes.  Since FUSE-T requires no kernel extensions, it is easier to maintain for MacOS's constantly changing kernel space, and that one has to go into recovery mode to select a lower security option for FUSE, while FUSE-T doesn't require this.

# Additional information

FUSE-T's website:  https://www.fuse-t.org/

# Your Environment

Please tell us more about your environment

**VeraCrypt version:** VeraCrypt 1.25.9

**Operating system and version:** macOS 13.3.1

**System type:** arm64

**Comment by ZelnickB** [comment]
+1 to this issue. FUSE-T markets itself as a "Drop-in replacement for osxfuse," so even though I haven't looked at the VeraCrypt code, I can't imagine that the implementation would be very difficult.

Since VeraCrypt aims to be transparent, I also think that it's problematic that both MacFUSE and FUSE-T are closed-source programs, but I suppose that's a different discussion.

**Comment by arifzeeshan** [comment]
Well,  it will work with minor changes to veracrypt, the developer of Fuse-T did compile and [tested](https://github.com/macos-fuse-t/fuse-t/issues/11#issuecomment-1320898562) it with minor changes.

**Comment by idrassi** [comment]
It is indeed becoming harder to configure macFUSE on recent macOS versions for non-technical users. This limits the userbase of VeraCrypt and supporting Fuse-T is a solution to this problem.

Thank you @arifzeeshan for sharing the confirmation of Fuse-T author that minimal changes are needed to make it compatible. He didn't share these changes but I give it a try in the coming weeks and see how it goes.

**Comment by arifzeeshan** [comment]
Looking forward to the changes and since Fuse-T runs on a userspace, hope we will also have a portable version too.

**Comment by shadofficial** [comment]
Could you please provide an update on the progress made in this area? Any details on the specific changes implemented or the overall experience in integrating Fuse-T with VeraCrypt would be greatly appreciated. Thank you again for your efforts and for sharing this information!

**Comment by ZelnickB** [comment]
@idrassi Do you have a progress update on this feature?

**Comment by josephazrak** [comment]
Given the likely sensitivity of the data macOS VeraCrypt systems handle, it makes me deeply uncomfortable that it requires the user to inject closed-source kernel code written by a single developer. Actually, this concept would make me uncomfortable regardless of the data on the system—I don't personally trust Benjamin Fleischer to run kernel code on my machine I'm not allowed to see.

For all intents and purposes, it is unwise to use VeraCrypt for macOS until an open-source FS layer is adopted.

**Comment by keresztg** [comment]
Seems that Fuse-t author recently posted his changes to VeraCrypt [here](https://github.com/macos-fuse-t/fuse-t/issues/11#issuecomment-1893850227).

**Comment by ZelnickB** [comment]
What steps are necessary before macos-fuse-t/VeraCrypt can be merged into the main VeraCrypt repository?

**Comment by fortiko** [comment]
@idrassi thank you for your great work. It looks as if the PR from the Fuse-T repo is rather straight forward. Could you at least merge it into a new feature branch so that it gets some exposure and can be tested by the brave folks here? Thank you!

**Comment by beatschubser** [comment]
Any update on this? Would be great to have this option, I don't want to mess with kernel extensions on my new M1 machine.

**Comment by ZelnickB** [comment]
@beatschubser I believe that this is being held up by #1276 right now—see [this comment](https://github.com/veracrypt/VeraCrypt/pull/1276#issuecomment-1962749876).

**Comment by idrassi** [comment]
I have finally merged https://github.com/veracrypt/VeraCrypt/pull/1276 and added some enhancements on top of it.
I’m now looking into fuse-t support. I have reviewed the changes proposed by the fuse-t author in https://github.com/macos-fuse-t/fuse-t/issues/11, and they indeed look simple.

I will implement the changes to create a fuse-t build and share it here for testing.

**Comment by idrassi** [comment]
I have added FUSE-T support to VeraCrypt in this commit: https://github.com/veracrypt/VeraCrypt/commit/edde1d45f6de3e71b585b92e0e06d7499735c238

There was an issue that caused volumes not to mount the first time they are used, and it took time to figure out the cause and the fix. It turns out that FUSE-T is slower in making the content of the mounted volume control file available compared to MacFUSE, so I had to add a waiting loop (250 ms for each loop) in order for the mounted content to be accessible. You can see the loop here: https://github.com/veracrypt/VeraCrypt/commit/edde1d45f6de3e71b585b92e0e06d7499735c238?diff=split&w=0#diff-c700f1c203f50e9d8f52601149b2406b97501ce983d57afa1ad5a61a22bfe3c5R303

Another issue is that FUSE-T libraries are built in such a way that they target only macOS Sonoma (14.0), while VeraCrypt targets macOS Monterey as the minimum version. I don't know if this was done on purpose and macOS Ventura and Monterey are not supported by FUSE-T. So for now, we get this warning during the linking of VeraCrypt:

```
ld: warning: dylib (/usr/local/lib/libfuse-t.dylib) was built for newer macOS version (14.0) than being linked (12.0)
ld: warning: dylib (/usr/local/lib/libfuse-t.dylib) was built for newer macOS version (14.0) than being linked (12.0)
```

It would be better if FUSE-T could target Monterey or Ventura as the minimum version.

Apart from this, it seems to work well and it is certainly less painful than MacFUSE which requires disabling security settings on Apple Silicon.

The next step is to notarize the VeraCrypt binary for FUSE-T and share it here for testing.

**Comment by idrassi** [comment]
As promised, here is the notarized version of VeraCrypt that uses FUSE-T instead of MacFUSE:
[VeraCrypt_FUSE-T_1.26.11.dmg.zip](https://github.com/user-attachments/files/15944284/VeraCrypt_FUSE-T_1.26.11.dmg.zip)

SHA256 SUM:
`a2ee0316c409d78d721b381cb0610ae33dc6246c0497178780310836644b7b97  VeraCrypt_FUSE-T_1.26.11.dmg`

Based my tests, it seems to work well. Feedback is welcomed.

**Comment by pashynskykh** [comment]
Hi @idrassi,

First of all, thank you very much for your work! 

The new version seems to work on my M1 MacBook, but I found two (minor?) problems.

1. After creating a new volume and mounting it for the first time, I get the following error:
<img width="315" alt="Bildschirmfoto 2024-06-23 um 14 25 38" src="https://github.com/veracrypt/VeraCrypt/assets/14852335/993e8385-f3f5-429f-9648-48fb4fabeacb">

The next time it was mounted without any problems.

2. It always mounts to "/Voulmes/NO NAME". This may be related to a previous error and may be a problem in the future. I'd suggest using the voulme name in the mount point, e.g. "/Voulmes/my-vc-test-volume".

Best regards,
Vladyslav

**Comment by idrassi** [comment]
Thank you @pashynskykh for the quick feedback.

For the first issue, it is caused by a FUSE-T issue for which I have already implemented a workaround (a waiting loop of max 1.2 seconds): https://github.com/veracrypt/VeraCrypt/blob/210827d27cdab8ffe33e71653ff269cd3c02dcff/src/Core/Unix/CoreUnix.cpp#L306. On my M1 system, it is enough but it looks like it is not the case everywhere. So I will increase it, maybe to 5 or 10 seconds.

For the second issue, you certainly selected FAT as your filesystem in the wizard, and that’s why you are getting “NO NAME” in the mount point. This is hardcoded in VeraCrypt for FAT formatted volumes: https://github.com/veracrypt/VeraCrypt/blob/210827d27cdab8ffe33e71653ff269cd3c02dcff/src/Core/FatFormatter.cpp#L285. With other filesystems, you will get the standard “Untitled” name variation because VeraCrypt doesn’t set any volume name explicitly.

You can always change the volume name to your liking in Finder by selecting “Rename.” Once done, you will start to see this customized name when you mount the volume with VeraCrypt.

**Comment by pashynskykh** [comment]
@idrassi thanks for the detailed answer. Yes, for the second "problem" I used FAT. It really does make sense to avoid using any names at all, as the name could theoretically be a privacy/security issue if not chosen carefully. Thanks for the clarification.

I'll have more time to play with the app tomorrow. If I find anything else, I'll post an update here.

Thanks again for all your work. It's really great to see progress on this issue, as FUSE-T is a superior choice to macFUSE.

Best regards,
Vladyslav

**Comment by tarbs** [comment]
The new 1.26.11 is working well for me (M1 Max MacBook Pro/MacOS 14.5 + external USB C SSD).  

I haven't tested extensively but everything is good so far.

Andy

**Comment by idrassi** [comment]
Thank you @tarbs for the feedback. This shows that the current workaround for the FUSE-T delay issue does work on some machines other than mine.

I’m starting to like FUSE-T and I hope further tests by the community will not uncover any major issues because it is a good candidate for replacing MacFUSE as the default FUSE layer, esopecially that it is easier to deploy on Apple Silicon for non-technical users.

**Comment by idrassi** [comment]
@pashynskykh, I have increased the waiting time for FUSE-T to 5 seconds and prepared a new build with this change. Please let me know if this solves the problem of mounting newly created volumes.

[VeraCrypt_FUSE-T_1.26.12.dmg.zip](https://github.com/user-attachments/files/15945940/VeraCrypt_FUSE-T_1.26.12.dmg.zip)

SHA256SUM:
`bc7f2a37560e46d67b42cf699b29871ac376f8551a10d8bc01f50a80265f6cae  Setup/MacOSX/VeraCrypt_FUSE-T_1.26.12.dmg`

**Comment by pashynskykh** [comment]
@idrassi just checked: created a new volume and mounted it. I didn't get any errors this time. Thank you very much!

**Comment by chimbori** [comment]
Just tested on a corp machine with kernel extensions locked down, and thank you! This works perfectly!

**Comment by pashynskykh** [comment]
I have been playing more with 1.26.12 today and have not seen any problems. I am happy that it can finally be used on my Macbook with FUSE-T. Thanks a lot @idrassi!

**Comment by idrassi** [comment]
Thank you all for the testing, specially @pashynskykh. Things look good for FUSE-T support. It will be the recommended version of VeraCrypt for Apple Silicon.

**Comment by kapitainsky** [comment]
It works in regards to using fuse-t - fantastic news.

But I have noticed that it has missing "TrueCrypt Mode" when mounting. Any particular reason to remove it? I have still some old TrueCrypt disks I mount from time to time.

**Comment by idrassi** [comment]
@kapitainsky Thank you for the feedback.

Concerning TrueCrypt support, it was indeed dropped in version 1.26.7 in order to simplify the code base and add room for new Blake2s algorithm. For users like you that have old TrueCrypt volume, it is possible to convert them to VeraCrypt format without loosing any data as indicated in the documentation: https://veracrypt.fr/en/Converting%20TrueCrypt%20volumes%20and%20partitions.html
You will have to use version 1.25.9 for this.

**Comment by kapitainsky** [comment]
Maybe good excuse to convert these old volumes:) 

BTW. I am testing VC and FUSE-T on Intel mac - so far all works perfectly.

**Comment by larelogio** [comment]
Installed in a fresh env (first VC installation), Sonoma (24.5) Mac Book Air M3.

FUSE-T 1.0.38 (sshfs: 1.0.2).

First tests OK. Will test on a old volume as soon as possible.

> As promised, here is the notarized version of VeraCrypt that uses FUSE-T instead of MacFUSE: [VeraCrypt_FUSE-T_1.26.11.dmg.zip](https://github.com/user-attachments/files/15944284/VeraCrypt_FUSE-T_1.26.11.dmg.zip)
> 
> SHA256 SUM: `a2ee0316c409d78d721b381cb0610ae33dc6246c0497178780310836644b7b97 VeraCrypt_FUSE-T_1.26.11.dmg`
> 
> Based my tests, it seems to work well. Feedback is welcomed.

**Comment by theprivacydesigner** [comment]
@idrassi Thank you for updating fuse-t.
Is the code posted somewhere?
I can’t seem to find it.

**Comment by idrassi** [comment]
you can grab the latest code from master, the fuse-t changes will be there.

**Comment by theprivacydesigner** [comment]
PSA: **fuse-t is NOT open source**.

I wrongly assumed that fuse-t’s GitHub had code in it.
There’s no code on it.

See for yourself [how the fuse-t author responds to requests to open source the project](https://github.com/macos-fuse-t/fuse-t/issues/1).

**Comment by idrassi** [comment]
I misunderstood your query:I thought you were asking about VeraCrypt changes related to fuse-t.

Regarding fuse-t not publishing its source code, the same applies to MacFUSE, which VeraCrypt has always used (although it was open source at some point in the past).

I wish MacFUSE and fuse-t were open source but I also understand why their authors have chosen not to do so. Many people simply copy their source code to create products without donating to the projects or funding the authors. These two projects are amazing and each is developed by a single person who needs some revenue to sustain their work. By not publishing the source code, they can sell commercial licenses and provide professional support, which is a good way to balance offering a free product with maintaining a decent revenue stream.

I can personally relate to this: the lack of revenue from VeraCrypt prevents me from working on it full-time. I often have to pause development to focus on my personal life and day job.

**Comment by ron4976** [comment]
I cant get this to work. Not sure what i am doing wrong, but it it says can't mount volume. Then it displays as if its there, but cant access it. Can't dismount it as it says Resource Busy. I have Veracrypt Fuse-t 1.26.14, MacOs Sequoia 15.0. Any help would be greatly appreciated.

**Comment by idrassi** [comment]
@ron4976  You didn’t share the error message you encountered but I suspect it might be the "hdiutil: attach failed - no mountable file systems" error. Can you confirm? If so, this is related to the exFAT issue affecting macOS Sequoia, for which there is already an open issue: https://github.com/veracrypt/VeraCrypt/issues/1422.

In short, this is a macOS issue, and currently, the only workaround is to create a new exFAT volume on Sequoia and copy the data from the old volume to the new one. Users have shared more information about this in the linked issue.

**Comment by ron4976** [comment]
@idrassi I did read this before, and used your advice to use Fuse-T veracrypt. But still have the same issue. So, what you say here is a i need to create a new exFAT volume.

**Comment by idrassi** [comment]
@ron4976 Yes, because Sequoia introduced an incompatibility issue with the exFAT filesystem, which requires creating a new exFAT volume on Sequoia.

Apple should be looking into this, as Sequoia is the only operating system experiencing this issue, unlike previous versions of macOS or other operating systems (Windows, Linux..). However, I have no way to report this to Apple and I'm not sure if they would even care since they have their own filesystem, and exFAT is just a third-party filesystem for them.

**Comment by idrassi** [comment]
@ron4976 Confirmed: Apple really doesn't care: https://discussions.apple.com/thread/255188289?sortBy=rank. This issue has been reported to them since last year, and there's still no fix from their side.

**Comment by RonLesser** [comment]
I've long used Veracrypt on Windows but am moving to a Mac...
This is probably a dumb question, but is Fuse-T included in the Veracrypt build or do I install Fuse-T first, and then install the Fuse-T version of Veracrypt?
Thanks.

**Comment by idrassi** [comment]
@RonLesser : FUSE-T is not included in VeraCrypt installation package, so you will have to download it from https://www.fuse-t.org/ and install it before installing VeraCrypt.

**Comment by RonLesser** [comment]
Thanks.

**Comment by harrison-labs** [comment]
How is FUSE-T functioning on Apple Silicon macs at this point?  I've heard from some people that they've into corruption issues and others say it's working just fine

**Comment by Inkoming** [comment]
Using brew to install VeraCrypt but it looks like I can't use it anymore on macOS... I do not want to use macfuse anymore. FUSE-T is the future. Cryptomator already added it and it's working smooth.

```
==> Installing dependencies: macfuse
==> Caveats
macfuse requires a kernel extension to work.
```

@idrassi I love your work and the soft, I know you're working on it on your spare time. Do you have any ETA about when will you be supporting FUSET by default please? Thanks so much!

**Comment by kapitainsky** [comment]
@Inkoming @chimbori  - not sure what you mean. Every release since v1.26.14 includes version with fuse-t support, e.g. the latest one:

<img width="1188" alt="Image" src="https://github.com/user-attachments/assets/c584d79d-2806-4515-881c-bd4d73ba11d2" />

Regarding brew version - you have to talk to brew veracrypt cask maintainer. Or DYI. What is needed is new cask with fuse-t version. Old one should not be changed as it would break for people using macFUSE.

**Comment by Inkoming** [comment]
Thanks a lot, excellent arguments here, will do it this way.

**Comment by idrassi** [comment]
As indicated by @kapitainsky, brew is a third-party package manager that is outside of our control, so it must be used with caution. We recommend direct install using official packages.

Since FUSE-T support has been implemented, I will close this issue as done. For problems affecting FUSE-T integration, please open new issue.

Big thanks to all participants in this long thread.

**Comment by chimbori** [comment]
@idrassi Sorry I missed the official builds, agreed I’ll contact the Homebrew maintainers.

Super awesome thanks for working on this, we all appreciate it!

**Comment by echuber2** [comment]
It sounds like FUSE-T has implemented an SMB backend now, which is supposed to be faster (based on posts on their GitHub issues). It would be nice if VeraCrypt could switch between the NFS an SMB backends.
